python/ray/serve/application.py

```diff
@@ -0,0 +1,190 @@
+from typing import List, Dict, TextIO, Optional, Union
+import json
+import yaml
+import time
+
+import ray
+from ray import serve
+from ray.serve.api import (
+    Deployment,
+    _get_global_client,
+)
+from ray.serve.common import DeploymentStatus
+from ray.dashboard.modules.serve.schema import (
+    ServeApplicationSchema,
+    serve_application_to_schema,
+    schema_to_serve_application,
+    serve_application_status_to_schema,
+)
+from ray.dashboard.modules.serve.sdk import ServeSubmissionClient
+from ray.serve.utils import logger
+
+
+class Application:
+    def __init__(self, deployments: List[Deployment]):
+        self._deployments = dict()
+        for idx, d in enumerate(deployments):
+            if isinstance(d, Deployment):
+                self._deployments[d.name] = d
+            else:
+                raise TypeError(
+                    f"Got object of type {type(d)} at index "
+                    f"{idx}. Expected only Deployment objects to "
+                    "be passed in."
+                )
+
+    def add_deployment(self, deployment: Deployment):
+        """Add deployment to the list. Validate name uniqueness."""
+
+        if not isinstance(deployment, Deployment):
+            raise TypeError(f"Got {type(deployment)}. Expected deployment.")
+
+        self._deployments[deployment.name] = deployment
+
```

Comment:
we can also have `set_options` function that configures options for each deployment via python, this also leads to details about mutability of `Deployment` objects in Application class.  Our current public API is 

"
Deployments should be created, retrieved, and updated using
        `@serve.deployment`, `serve.get_deployment`, and `Deployment.options`,
        respectively.
"

Reply:
I added `__setattr__` just now to do something similar. If you want to update a deployment's options in an Application object, you should be able to do:

```
app.deployment_name = app.deployment_name.options(...)
```

or

```
app[deployment_name] = app[deployment_name].options(...)
```

How does that look?

Reply:
hmm i remember ed and simon likes `set_options` from review last week so it's like `app.deployment_name.set_options(num_replicas=2)` so we appears to be setting a field on an existing deployment of an app, rather than creating and re-assigning to it. (Under the hood we're doing the same tho)

I'm leaning towards `set_options` but don't have super strong opinion on this, given I didn't put up the public `.options()` API in existing codebase. So you might want to check with them to see if they're strongly opinionated.

Reply:
Hmm I do think `set_options` is nicer in this case, but in favor of minimizing changes let's go with what shreyas described above for now:
`app.ingress = app.ingress.options(num_replicas=100)`